File: ARC/ARC037/b.py
import sys
sys.setrecursionlimit(10**9)

N, M = map(int, input().split(" "))
edge = dict()
for _ in range(M):
    u, v = map(int, input().split(" "))
    if str(u) in edge:
        edge[str(u)].add(v)
    else:
        edge[str(u)] = set({v})
    if str(v) in edge:
        edge[str(v)].add(u)
    else:
        edge[str(v)] = set({u})

visited = set()

def dfs(x, x0=None):
    if {x}.issubset(visited):
        return False
    else:
        visited.add(x)
        time = 0
        same = 0
        if str(x) not in edge:
            return True
        for y in edge[str(x)]:
            if y != x0:
                if dfs(y, x) == True:
                    time += 1
            else:
                same += 1
        if time == len(edge[str(x)]) - same:
            return True
        else:
            return False

tree = 0
for i in range(N):
    if {i+1}.issubset(visited):
        pass
    else:
        if dfs(i+1) == True:
            tree += 1

print(tree)

File: ARC/ARC037/b.cpp
#include <iostream>
using namespace std;

int N, M;
int edge[101][101] = {0};
int visited[101] = {0};

bool dfs(int x, int x0) {
    if (visited[x] == 1) return false;
    visited[x] = 1;
    for (int itr=1; itr<=N; ++itr) {
        if (itr != x0 && edge[x][itr] == 1) {
            if (!dfs(itr, x)) return false;
        }
    }
    return true;
}

int main() {
    cin >> N >> M;
    for (int i=0; i<M; ++i) {
        int u, v;
        cin >> u >> v;
        edge[u][v] = 1;
        edge[v][u] = 1;
    }
    
    int ans = 0;
    for (int i=1; i<=N; ++i) {
        if (visited[i] == 0) {
            if (dfs(i, -1)) ans += 1;
        }
    }

    cout << ans << endl;

    return 0;
}